addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/environment-resource-service.js

```diff
@@ -0,0 +1,287 @@
+   * @param studies an array of StudyEntity that are associated with this env. IMPORTANT: each element
+   * in the array is the standard StudyEntity, however, there is one additional attributes added to each
+   * of the StudyEntity. This additional attribute is called 'envPermission', it is an object with the
+   * following shape: { read: true/false, write: true/false }
+   */
+  async deallocateStudyResources(requestContext, { environmentScEntity, studies: allStudies, memberAccountId }) {
+    // We are only interested in studies with bucket access = 'roles'
+    const studies = _.filter(allStudies, study => study.bucketAccess === 'roles');
+
+    if (_.isEmpty(studies)) return; // No relevant studies for this access strategy, so we are done
+
+    // The logic
+    // - We group studies by their appRoleArn, this makes the locking a bit more efficient
+    // - Then for each group (in batches), we do the following logic:
+    //   - For each study, we ask the filesystem role system service to deallocate the role for us
+    //   - We deleted the study role entry in the studies map
+    // - After we release all the necessary filesystem roles, we update the study roles in the environmentScEntity
+
+    const studyRoles = environmentScEntity.studyRoles || {};
+    const lockService = await this.service('lockService');
+    const filesystemRoleService = await this.service('roles-only/filesystemRoleService');
+    const environmentService = await this.service('environmentScService');
+
+    // For certain calls, the system context is used instead of the request context
+    const systemContext = getSystemRequestContext();
+
+    // groups is an object where the keys are the appRoleArn and the values are arrays
+    // of the studyEntities. Example:
+    // groups = { '<appRoleArn1>': [ <studyEntity>, ... ], '<appRoleArn1>': [ <studyEntity>, ... ]}
+    const groups = _.groupBy(studies, 'appRoleArn');
+    const groupProcessor = async group => {
+      // group is an array of study entities where they all have the same app role arn
+      const appRoleArn = _.get(_.nth(group, 0), 'appRoleArn', '');
+      const lockId = `roles-only-access-app-role-${appRoleArn}`;
+
+      await lockService.tryWriteLockAndRun({ id: lockId }, async () => {
+        // For each study, we ask the filesystem role system service to deallocate the role for us
+        // eslint-disable-next-line no-restricted-syntax
+        for (const study of group) {
+          const fsRoleArn = studyRoles[study.id];
+          // eslint-disable-next-line no-continue
+          if (_.isEmpty(fsRoleArn)) continue;
+          await filesystemRoleService.deallocateRole(
+            systemContext,
+            fsRoleArn,
+            study,
+            environmentScEntity,
+            memberAccountId,
+          );
+
+          delete studyRoles[study.id];
+        }
+      });
+    };
```

Comment:
nit: Is it worth moving this code into a shared function? I feel its fine as it is but wondering if you thought about it?

Reply:
Didn't really think about putting that in a function. Thinking about it now, I feel it is okay the way it is for now.  Maybe something I can address if we add more logic to this exact loop.